src/main/flight/pid.c

```diff
@@ -287,23 +293,76 @@ void pidResetIterm(void)
     }
 }
 
+
 #ifdef USE_WING
-static float getWingTpaArgument(float throttle)
+static float calcWingThrottle(void)
+{
+    float batteryThrottleFactor = 1.0f;
+    if (pidRuntime.tpaSpeed.maxVoltage > 0.0f) {
+        batteryThrottleFactor = getBatteryVoltageLatest() / 100.0f / pidRuntime.tpaSpeed.maxVoltage;
+        batteryThrottleFactor = constrainf(batteryThrottleFactor, 0.0f, 1.0f);
+    }
+
+    return getAverageMotorOutput() * batteryThrottleFactor;
+}
+
+
+static float calcWingAccelerationAdvanced(float throttle, float pitchAngle)
+{
+    const tpaSpeedParams_t *tpa = &pidRuntime.tpaSpeed;
+
+    const float thrust = (throttle * throttle - throttle * tpa->speed / tpa->propMaxSpeed) * tpa->twr * G_ACCELERATION;
+    const float drag = tpa->speed * tpa->speed / tpa->massDragRatio;
+    const float gravity = G_ACCELERATION * sin_approx(pitchAngle);
+
+    return thrust - drag + gravity;
+}
+
+
+static float calcWingAccelerationBasic(float throttle, float pitchAngle)
+{
+    const tpaSpeedParams_t *tpa = &pidRuntime.tpaSpeed;
+
+    const float thrust = throttle * throttle * tpa->twr * G_ACCELERATION;
+    const float drag = tpa->speed * tpa->speed / tpa->massDragRatio;
+    const float gravity = G_ACCELERATION * sin_approx(pitchAngle);
```

Comment:
(technically, you can precompute 'forward' vector from pitchOffset and do the matrix multiplication [quite simple, with all zeroes]. It will avoid $sin(sin^{-1}(x))$ and also avoid $sin^{-1}(sin(x))$ ambiguity (https://www.wolframalpha.com/input?i=plot+sin%5E-1%28sin%28x%29%29+) - IMO the problem is that you can't tell if flying upright or inverted - sin(pitch) is the same, gravity force is fine, but pitchOffset is added or subtracted from angle)

Reply:
i see the problem... How about something like that:
`const float gravity = G_ACCELERATION * sin_approx(pitch + cos(roll) * pitchOffset);`?